nft/src/main.sw

```diff
@@ -0,0 +1,474 @@
+        storage.owners.get(identity)
+    }
+
+    /// Returns the total supply for the NFT contract
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The NFT contract has not been initalized
+    fn get_total_supply() -> u64 {
+        require(storage.state != 0, InitError::NFTNotInitalized);
+        storage.token_supply
+    }
+
+    /// Returns whether the address is approved for all tokens
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The NFT contract has not been initalized
+    fn is_approved_for_all(owner: Identity, operator: Identity) -> bool {
+        require(storage.state != 0, InitError::NFTNotInitalized);
+
+        let hash: b256 = get_owner_operator_hash(owner, operator);
+        storage.operator_approval.get(hash)
+    }
+
+    /// Mints an NFT
+    ///
+    /// # Panics
+    /// 
+    /// The function will panic when:
+    /// - The NFT contract has not been initalized 
+    /// - The amount is set to 0
+    /// - More NFTs than supply is minted
+    /// - The to Identity is not valid
+    /// - The sender is not approved to mint
+    /// - The sender sent the wrong asset
+    /// - The sender did not pay enough tokens
+    fn mint(to: Identity, amount: u64) -> bool {
+        require(storage.state != 0, InitError::NFTNotInitalized);
+        require(amount != 0, InputError::MintAmountCannotBeZero);
+        require(
+            storage.token_supply >= (storage.token_count + amount), 
+            InputError::NotEnoughTokensToMint
+        );
+        require_identity_is_valid(to);
+
+        let sender: Identity = unwrap_identity(msg_sender());
+
+        /// Ensure that the sender is on the approved mint list
+        require(
+            !storage.access_control || 
+            (storage.access_control && storage.allowed_minters.get(sender)), 
+            AccessError::SenderDoesNotHaveAccessControl
+        );
+
+        /// Mint as many tokens as the sender has paid for
+        let mut i = 0;
+        while i < amount {
+            /// Create a new token id in sequential order
+            let token_id: u64 = storage.token_count + 1;
+            let empty_identity: Identity = Identity::Address(~Address::from(NATIVE_ASSET_ID));
+
+            /// Create the metadata for this new token with the owner 
+            let meta_data: MetaData = MetaData {
+                owner: to, approved: empty_identity
+            };
+            storage.meta_data.insert(token_id, meta_data);
+            storage.owners.insert(to, token_id);
+            
+            /// Increase the balance of the new owner
+            let mut balance = storage.balances.get(to);
+            storage.balances.insert(to, balance + 1);
+
+            /// Increment the token count and the number of tokens minted in this transaction
+            storage.token_count = storage.token_count + 1;
+            i = i + 1;
+
+            log(MintEvent{owner: to, token_id});
```

Comment:
Perhaps this should have some logic to keep track of how many are minted and only emit a log once at the end. Logging multiple times is not a good idea.